pkg/dataplane/client/xds.go

```diff
@@ -0,0 +1,76 @@
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/clusterlink-net/clusterlink/pkg/dataplane/server"
+)
+
+// XDSClient implements the client which fetches clusters and listeners
+type XDSClient struct {
+	dataplane          *server.Dataplane
+	controlplaneTarget string
+	tlsConfig          *tls.Config
+	fetchers           []*fetcher
+	errors             []error
+	logger             *logrus.Entry
+}
+
+func (x *XDSClient) runFetcher(_ *fetcher, resourceType string) error {
+	for {
+		time.Sleep(5 * time.Second)
+		conn, err := grpc.Dial(x.controlplaneTarget, grpc.WithTransportCredentials(credentials.NewTLS(x.tlsConfig)))
+		if err != nil {
+			x.logger.Errorf("Failed to dial controlplane xDS server: %v.", err)
+			continue
+		}
+		x.logger.Infof("Successfully connected to the controlplane xDS server.")
+
+		f, err := newFetcher(context.Background(), conn, resourceType, x.dataplane)
```

Comment:
@orozery suggested to use separate graph connections for both fetchers. It tries to create a connection again only upon failure of the fetcher. Otherwise, at a certain point there are two active connections only.